### PR TITLE
Opening hours widget

### PR DESCRIPTION
**Pull request by hstandaert**
### `Opening Hours` collection
UI component with variants for day/week/month representation of the opening hours. This collection consists of multiple widgets used to display the opening hours

#### Day widget
Displays the opening hours for 1 day. Has 3 variants:
- 1 channel (open)
- Multiple channels (open)
- Closed

#### Week Month widget
Tabbed rendition of the opening hours. On one hand for 7 days (starting on the current day) and on the other for the entire month (where the user can select other months as well)

### `Opening Hours Accordion` component
An extensive representation op the opening hours for all channels of a certain service (multiple widget)

### Included changes
- Empty DOM wrapper
- Theming

### What's not included
- Data fetching from the opening hours API
- DOM changes based on fetched data (for example the number of channels).

## Motivation and Context

In preparation of the contact pages for [digipolisgent/drupal8_site_stadgent](https://github.com/digipolisgent/drupal8_site_stadgent)

## Screenshots (if appropriate):
Designs for the page can be found at:

### Desktop

* Overview: https://app.zeplin.io/project/5a7973af3882dd06217f2015/screen/5bae207c605692661a5f977d
* Detail: https://app.zeplin.io/project/5a7973af3882dd06217f2015/screen/5bb224161f9ba41e7c5f9dbb
* Detail opening hours for this month: https://app.zeplin.io/project/5a7973af3882dd06217f2015/screen/5bae1f60e93a6803484d66e0
* Multi-channel detail (collapsed): https://app.zeplin.io/project/5a7973af3882dd06217f2015/screen/5bae1f61b734189ec039e39e
* Multi-channel detail (expanded):  https://app.zeplin.io/project/5a7973af3882dd06217f2015/screen/5bae1f6291a59a04653e0e52
* Infobox: https://app.zeplin.io/project/5a7973af3882dd06217f2015/screen/5bae1f61970d790789a80bd1

### Mobile

* Overview: https://app.zeplin.io/project/5a7973af3882dd06217f2015/screen/5bae1f640fa6cb780ce9f8b2
* Detail: https://app.zeplin.io/project/5a7973af3882dd06217f2015/screen/5bb22462d6cb254bd935b392
* Detail opening hours for this month: https://app.zeplin.io/project/5a7973af3882dd06217f2015/screen/5bae1f6379e49f0835d0d2be
* Multi-channel detail (collapsed): https://app.zeplin.io/project/5a7973af3882dd06217f2015/screen/5bae1f647044c49e8148e8cd
* Multi-channel detail (expanded):  https://app.zeplin.io/project/5a7973af3882dd06217f2015/screen/5bae1f62504c58535ac276a8
* Infobox: https://app.zeplin.io/project/5a7973af3882dd06217f2015/screen/5bae1f6391a59a04653e0f73

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [ ] I have updated the documentation accordingly.
- [ ] I have updated the style guide CHANGELOG accordingly.
- [ ] I have updated the gent_base CHANGELOG accordingly.
- [x] I have ran gulp axe on the compiled files and found no errors.

### CODE REVIEW

**Comment by hstandaert** [comment]
@delrueba, could you provide me with some credentials for http://openingsuren.web.test.gentgrp.gent.be so that I can test different use cases?

**Comment by hstandaert** [comment]
@delrueba, I've had a brief discussion with @pj-claeys and @lennartvava about the fact that functionality is being moved into the styleguide. 

Maybe we should re-think the decision to fetch data in the styleguide. This might transcend the goal of a styleguide, having stateless components rather than full-blown components that manipulate the DOM etc. 

wdyt?

**Comment by delrueba** [comment]
What we do in styleguide is describe desired functionality and provide a default script/library to achieve it.
(cfr. lightbox, tabs, modals, masonry...)
I see no difference in initiating a lightbox by providing certain markup vs initiating a ghent openinghours widget.

Remember there are also angular applications that will soon make use of these widgets... and they cannot use the drupal module for that...

If not in styleguide... we could move the script to openinghours itself and have it act as a CDN?

**Comment by hstandaert** [comment]
The examples you name are all UI/UX optimisations that require javascript, but data-fetching is a whole other level if you ask me. Just because we need shared functionality between applications doesn't mean that the styleguide is the right place for that. 

I'm a proponent for the CDN. Or maybe we can publish an npm package somewhere?

**Comment by delrueba** [comment]
So am I,
it's the html output of openinghours that determines how the widget works so it's only natural that it's also responsible for providing the script that adds function to the widget. You can still choose to use a custom implementation.

But, fetching the data itself... that looks like a frontend job to me since it's not prerendered on the server and the openinghours app is not responsible for knowing where you'd like to place all widgets. In particular the pages where we want to display multiple output for multiple channels.

**Comment by hstandaert** [comment]
Soooo, let's provide a CDN script on openinghours that can be loaded from any project and provides methods to fetch data.

That way : 
* The CDN script will be responsible for fetching data
* Styleguide will be responsible for adding JS events and UX/UI/accessibility optimisations

The 1 question that remains for me: do we really need a custom drupal widget then? (which is the only responsibility that is left for the drupal module, afaik?)

**Comment by delrueba** [comment]
Isn't it the drupal module that renders the html elements with serviceID / channelID?
I'd move the JS events for the month widget to the CDN too, compare it with the zoom buttons in a map.
But describe the desired functionality in styleguide so people can create their custom implementation.

Question that remains for me is:
how will we add multiple widgets for multiple channels on a page if backend doens't provide us with channel ID's?
Openinghours can't be made responsible to inject all those widgets in specific places on the page (cfr in tabs, under accordions...)

You'd still need to
- fetch all channels for a given service ID
- add titles, tabs, accordions for each channel...
- render DOM elements with channelID and serviceID
- run the openinghours script to fetch all widgets

**Comment by daften** [comment]
It is very simple: you need separation of concerns.
So fetching data is NOT part of the styleguide. The application itself decides how it wants to fetch data. Be it Ajax calls, backend calls so the data is cached in a reverse proxy, ... That logic goes in a separate module/plugin: a drupal module, an npm package, ...

The styleguide ONLY does rendering related stuff.

**Comment by hstandaert** [comment]
I'll rephrase my last comment then: 😀
We'll:
* Provide a CDN script in the opening-hours codebase that acts as a javascript API wrapper.
* Add DOM, DOM manipulations, JS events and accessibility optimisations in styleguide. ⚠️We'll add the CDN script as a JS dependency

Question still stands: the drupal module becomes redundant since we'll fetch the data based on an HTML attribute that is picked up by the styleguide scripts. Seems like overkill to foresee a custom backend widget for this..?

@daften, wdyt?

**Comment by daften** [comment]
Tagging @Jelle-S   about the idea of the JS CDN script in openinghours.
The Drupal module is redundant if front-end fetching is used, not when backend fetching is done. And I don't fully agree we should use the fetching data script as a dependency. We can document it, but it's up to the framework to use it. If a Vue/Angular application wants to code fetching of data itself, they should be able to IMO. Or if an application uses backend fetching to store the openinghours in the page so reverse proxy caching can be done on that resource, it should be possible.

**Comment by Jelle-S** [comment]
CDN or npm package both sound good to me, personally, I wouldn't include it in the styleguide.

In my mind npm package sounds better than cdn, but that's not my forte, so I won't push either way.

Regarding the Drupal module: it does not become redundant since it provides a field that can be used on an entity where users/editors can select the service and channel to display the opening hours for. That is something Drupal-specific that cannot be included in a styleguide or JS library. The JS code in that module will probably become reduntant. We might still need some JS code in that module to have a Drupal wrapper around the library.

File: styleguide/components/41-organisms/opening-hours-accordion/opening-hours-accordion.twig [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<div class="opening-hours-accordion">
+  <div class="opening-hours-accordion__item">
+    <h3>Appointments</h3>
+    {% include '@day-widget' with {
+      open: true
+    } %}
+
+    {% set openingHoursMonth %}
```
Comment (delrueba) [inline styleguide/components/41-organisms/opening-hours-accordion/opening-hours-accordion.twig:8]: :+1: very elegant!

File: styleguide/components/31-molecules/opening-hours/week-month-widget/opening-hours-month-widget.functions.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+        if (nextElem) {
+          e.preventDefault();
+          nextElem.click();
+        }
+      };
+
+      switch (keyCode) {
+        case 37: // previous (left arrow)
+          changeFocus(--currentPosition, 31, 30);
```
Comment (delrueba) [inline styleguide/components/31-molecules/opening-hours/week-month-widget/opening-hours-month-widget.functions.js]: missing 29 and 28?

Comment (delrueba) [inline styleguide/components/31-molecules/opening-hours/week-month-widget/opening-hours-month-widget.functions.js]:
```suggestion
          changeFocus(--currentPosition, 31, 30, 29, 28);
```

File: styleguide/components/31-molecules/opening-hours/week-month-widget/opening-hours-month-widget.functions.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    }
+    else {
+      root.OpeningHoursMonthWidget = factory();
+    }
+  }
+})(this || window, function () {
+  return function (elem, options) {
+    if (typeof gent_styleguide === 'undefined') {
+      // eslint-disable-next-line no-console
+      console.error('You need to include base.js.');
```
Comment (delrueba) [inline styleguide/components/31-molecules/opening-hours/week-month-widget/opening-hours-month-widget.functions.js]: Do we?

**Review by delrueba — CHANGES_REQUESTED** [review]

**Review by delrueba — APPROVED** [review]

**Pull request merged**